### PR TITLE
Added from_config classmethod for Starlite

### LINKED ISSUE #1120: Enhancement: Allow configuration of an application via an `AppConfig` instance
We already have `AppConfig` in place, so it would be nice to be able to do something like 

```python
from starlite import Starlite, AppConfig

config = AppConfig(...)

app = Starlite.from_config(config)
```

### PR DESCRIPTION
**Pull request by wassafshahzad**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.rst`?
- [ ] Have you got 100% test coverage on new code?
- [ ] Have you updated the prose documentation?
- [ ] Have you updated the reference documentation?

### Which ticket is this PR associated with ?
Closes #1120 

### What does this PR do
This PR adds the following features
- Adds a configuration object in Starlite constructor
- Adds sane defaults in AppConfig
- Updates pre-commits to fix isort issue

### Background Context
This PR was accidently closed. Follow previous discussion [here](https://github.com/starlite-api/starlite/pull/1173)

### CODE REVIEW

File: starlite/app.py [hunk 1]
```diff
@@ -401,22 +402,22 @@ def __init__(
...
-        for route_handler in config.route_handlers:
+        for route_handler in config.route_handlers:  # type: ignore
             self.register(route_handler)
 
         if self.debug and isinstance(self.logging_config, LoggingConfig):
             self.logging_config.loggers["starlite"]["level"] = "DEBUG"
 
         if self.logging_config:
-            self.get_logger = self.logging_config.configure()
+            self.get_logger = self.logging_config.configure()  # type: ignore
```
Comment (peterschutt) [inline starlite/app.py:412]: I think the ignore is necessary here as we allow the logging config to be both `None` or `Empty` but `Empty` doesn't evaluate falsey, so at this point the config object could be `Empty` or a config object. It that is the case, the type error is correct and we shouldn't be ignoring it.

File: starlite/config/app.py [hunk 2]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
+    guards: List[Guard] = []
     """A list of :class:`Guard <starlite.types.Guard>` callables."""
-    initial_state: InitialStateType
+    initial_state: Optional[InitialStateType] = None
     """An object from which to initialize the app state."""
-    listeners: List[EventListener]
+    listeners: List[EventListener] = []
     """A list of :class:`EventListener <starlite.events.listener.EventListener>`."""
-    logging_config: Optional[BaseLoggingConfig]
+    logging_config: Union["BaseLoggingConfig", "EmptyType", None]
```
Comment (peterschutt) [inline starlite/config/app.py:141]:
```suggestion
    logging_config: Optional[Union["BaseLoggingConfig", "EmptyType"]]
```

File: starlite/config/app.py [hunk 3]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
     """Defaults to :data:`DEFAULT_OPENAPI_CONFIG <starlite.app.DEFAULT_OPENAPI_CONFIG>`"""
-    opt: Dict[str, Any]
+    opt: Dict[str, Any] = {}
     """A string keyed dictionary of arbitrary values that can be accessed in :class:`Guards <starlite.types.Guard>` or
     wherever you have access to :class:`Request <starlite.connection.request.Request>` or :class:`ASGI Scope <starlite.types.Scope>`.
 
     Can be overridden by routers and router handlers.
     """
-    parameters: ParametersMap
+    parameters: Optional[ParametersMap] = None
```
Comment (peterschutt) [inline starlite/config/app.py:157]:
```suggestion
    parameters: ParametersMap = {}
```

File: starlite/config/app.py [hunk 4]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
     """A mapping of :class:`Parameter <starlite.params.Parameter>` definitions available to all application paths."""
-    plugins: List[PluginProtocol]
+    plugins: List[PluginProtocol] = []
     """List of :class:`SerializationPluginProtocol <starlite.plugins.base.SerializationPluginProtocol>`."""
     request_class: Optional[Type[Request]]
     """An optional subclass of :class:`Request <starlite.connection.request.Request>` to use for http connections."""
     response_class: Optional[ResponseType]
     """A custom subclass of [starlite.response.Response] to be used as the app's default response."""
-    response_cookies: ResponseCookies
+    response_cookies: Optional["ResponseCookies"] = None
```
Comment (peterschutt) [inline starlite/config/app.py:165]:
```suggestion
    response_cookies: ResponseCookies = []
```

File: starlite/config/app.py [hunk 5]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
     """A list of [Cookie](starlite.datastructures.Cookie] instances."""
-    response_headers: ResponseHeadersMap
+    response_headers: Optional["ResponseHeadersMap"] = None
```
Comment (peterschutt) [inline starlite/config/app.py:167]:
```suggestion
    response_headers: ResponseHeadersMap = {}
```

File: starlite/config/app.py [hunk 6]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
     """A string keyed dictionary mapping :class:`ResponseHeader <starlite.datastructures.ResponseHeader>` instances."""
-    route_handlers: List[ControllerRouterHandler]
+    route_handlers: "OptionalSequence[ControllerRouterHandler]" = None
```
Comment (peterschutt) [inline starlite/config/app.py:169]:
```suggestion
    route_handlers: List[ControllerRouterHandler] = []
```

File: starlite/config/app.py [hunk 7]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
     """A required list of route handlers, which can include instances of :class:`Router <starlite.router.Router>`, subclasses
     of.
 
     :class:`Controller <starlite.controller.Controller>` or any function decorated by the route handler decorators.
     """
-    security: List[SecurityRequirement]
+    security: "OptionalSequence[SecurityRequirement]" = []
```
Comment (peterschutt) [inline starlite/config/app.py:175]:
```suggestion
    security: List[SecurityRequirement] = []
```

File: starlite/config/app.py [hunk 8]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
...
     """A list of dictionaries that will be added to the schema of all route handlers in the application. See.
 
     :class:`SecurityRequirement <pydantic_openapi_schema.v3_1_0.security_requirement.SecurityRequirement>` for details.
     """
-    static_files_config: List[StaticFilesConfig]
+    static_files_config: "OptionalSequence[StaticFilesConfig]" = None
```
Comment (peterschutt) [inline starlite/config/app.py:180]:
```suggestion
    static_files_config: List[StaticFilesConfig] = []
```

**Review by peterschutt — CHANGES_REQUESTED** [review]
The type changes on the config object shouldn't be required and would be the source of the type issues that you've encountered when using the values from the config object in `Starlite.__init__()`

File: starlite/config/app.py (hunk 2 above)
Comment (peterschutt) [reply starlite/config/app.py:141]: I don't believe this isn't intended to be allowed to be both union `Empty` _and_ `Optional` on the config object. The possible `Empty` value is resolved away [here](https://github.com/starlite-api/starlite/blob/4cc867c0c5f49449b64932a391029b41b295b14c/starlite/app.py#L304).

File: starlite/app.py [hunk 9]
```diff
@@ -8,6 +8,7 @@
     List,
     Mapping,
     Optional,
+    Self,
```
[the remark this answers is not in the document]
Comment (wassafshahzad) [reply starlite/app.py:11]: > The type changes on the config object shouldn't be required and would be the source of the type issues that you've encountered when using the values from the config object in `Starlite.__init__()`

Thank you for informing me, I will revert those changes immediately, Any reason why those are required ? (I'm new to typing in python)

File: starlite/app.py [hunk 10]
```diff
@@ -401,22 +402,22 @@ def __init__(
         for plugin in (p for p in config.plugins if isinstance(p, InitPluginProtocol)):
             plugin.on_app_init(app=self)
 
-        for route_handler in config.route_handlers:
+        for route_handler in config.route_handlers:  # type: ignore
             self.register(route_handler)
```
[the remark this answers is not in the document]
Comment (wassafshahzad) [reply starlite/app.py:406]: I believe my changes in config objects types must have caused this issue

File: starlite/config/app.py [hunk 11]
```diff
@@ -116,67 +119,67 @@ class Config(BaseConfig):
     """If set this enables the builtin CORS middleware."""
     csrf_config: Optional[CSRFConfig]
     """If set this enables the builtin CSRF middleware."""
-    debug: bool
+    debug: bool = True
```
[the remark this answers is not in the document]
Comment (wassafshahzad) [reply starlite/config/app.py:122]: No clue about this, I thought debug by default would have been safer. I can set it to False

[the remark this answers is not in the document]
Comment (wassafshahzad) [reply starlite/config/app.py:122]: Sure, Will do,
BTW @provinzkraut, would you be able to guide me on @peterschutt [comment ](https://github.com/starlite-api/starlite/pull/1190#pullrequestreview-1294781156)

**Review by provinzkraut — APPROVED** [review]

**Review by JacobCoffee — APPROVED** [review]

**Comment by Goldziher** [comment]
we are still missing documentation for this, @provinzkraut - should this be done in this scope? if not we can merge.

**Comment by Goldziher** [comment]
@all-contributors add @wassafshahzad code

**Comment by wassafshahzad** [comment]
Should I wait for @peterschutt review ?

**Comment by provinzkraut** [comment]
> we are still missing documentation for this, @provinzkraut - should this be done in this scope? if not we can merge. 

I'll add this to the docs elsewhere

**Review by peterschutt — APPROVED** [review]

**Pull request merged**